clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -6191,8 +6181,28 @@ void CodeGenModule::EmitGlobalVarDefinition(const VarDecl *D,
         AddGlobalSYCLIRAttributes(GV, RD);
       // If VarDecl has a type decorated with SYCL device_global attribute 
       // emit IR attribute 'sycl-unique-id'.
-      if (RD->hasAttr<SYCLDeviceGlobalAttr>())
+      if (RD->hasAttr<SYCLDeviceGlobalAttr>()) {
         addSYCLUniqueID(GV, D, Context);
+        if (Linkage == llvm::GlobalValue::InternalLinkage) {
+          // Despite being `static`, static device globals need to be linked
+          // externally as symbols must persist across the host-device boundary
+          // and not removed or tampered with in later optimizations. Since we
+          // are linking these symbols externally we need a way to 'hide' the
+          // static device global symbols from other TUs that may have the same
+          // static symbols. We can 'hide' the symbols from other TUs despite
+          // being linked externally by prefixing the symbol with a UID string.
+          //
+          // Also prefix the symbol with __ so the symbol is not in user space.
+          auto builtinString =
+              "__" + SYCLUniqueStableIdExpr::ComputeName(Context, D);
+          GV->setName(builtinString);
```

Comment:
If we're doing this we should probably also frontend test these codegen changes. Check the IR for some kind of `__uid[A-Z0-9]device_global_var` name on a static `device_global` (or similar), and `@device_global_var` on non-static,const/extern/etc ones

Reply:
Yes tests to follow. Currently working on the case of two different kernels using the same dev global. Since the funcs are processed linearly the second func doesn't see the symbol renaming so trying to figure out a way to deal with this.

Reply:
The modified tests cover this